### PR TITLE
Replace Lucene DataInput/DataOutput with Elasticsearch StreamInput/StreamOutput when reading/writing geo_shape doc values

### PR DESCRIPTION
**Pull request by iverase**
Currently we are using Lucene abstractions DataInput / DataOutput to read / write geoshape doc values. This is problematic once we move to Lucene 9.0 as it has changed the endianness. This PR replaces this abstraction with Elasticsearch StreamInput / StreamOuput. In order to do that we have done the following replacements:

* Lucene ByteBuffersDataOutput has been replaced with Elasticsearch BytesStreamOutput.
* There is not Elasticsearch equivalent to Lucene ByteArrayDataInput. Therefore we have added a new class called ByteArrayStreamInput that is heavily based in Lucene implementation and use it as replacement.

We assume Elasticsearch StreamInput/StreamOutput is compatible with Lucene one so we can replace one with the other without worrying on backwards compatibility

### CODE REVIEW

**Comment by nik9000** [comment]
Drive by comment: I thought you could do something like `new
BytesArray(array).streamInput()` and get a stream.

**Comment by iverase** [comment]
@nik9000 The main two features of `ByteArrayStreamInput` are:

* Is reusable. You don't want to create anew Object for each document.
* random reads. You can set the position on the stream so you can go forwards and backwards.

Does this answer your question?

**Comment by nik9000** [comment]
> Does this answer your question?

Yup. Thanks!

**Comment by iverase** [comment]
@elasticmachine run elasticsearch-ci/docs

File: server/src/test/java/org/elasticsearch/common/io/stream/ByteBufferStreamInputTests.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ */
+
+package org.elasticsearch.common.io.stream;
+
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.bytes.BytesReference;
+
+import java.io.IOException;
+
+public class ByteBufferStreamInputTests extends AbstractStreamTests {
```
Comment (jpountz) [inline server/src/test/java/org/elasticsearch/common/io/stream/ByteBufferStreamInputTests.java:16]: It isn't really about the ByteBuffer class, more about BytesRef?

**Review by jpountz — APPROVED** [review]
This makes sense to me.

File: server/src/test/java/org/elasticsearch/common/io/stream/ByteBufferStreamInputTests.java (hunk 1 above)
Comment (iverase) [reply server/src/test/java/org/elasticsearch/common/io/stream/ByteBufferStreamInputTests.java:17]: I changed to so it is more obvious how the ByteBufferStreamInput is used in the test.

**Pull request merged**